# Лабораторная работа №3
## Выполнила студентка группы БФИ2001 Тимофеева Дарья Валерьевна

### Оглавление
1. [Задание 1](#Задание-№1)
2. [Задание 2](#Задание-№2)
4. [Вывод](#Вывод)

> Дополнительные модули, использованные при выполнение лабораторной

In [2]:
# Необходим при замере скорости выполнения кода
from datetime import datetime as t
# Нужен для создания словаря в алг. Бойера-Мура
from collections import defaultdict

### Задание №1
Реализовать методы поиска подстроки в строке. Добавить возможность ввода строки и подстроки с клавиатуры. Предусмотреть возможность существования пробела. Реализовать возможность выбора опции чувствительности или нечувствительности к регистру. Оценить время работы каждого алгоритма поиска и сравнить его со временем работы стандартной функции поиска, используемой в выбранном языке программирования.

#### Алгоритм Кнута-Морриса-Пратта

In [8]:
text = input("Cтрока: ")
sub_text = input("Подстрока: ")
button = input("Выбор опции чувствительности к регистру -  on or off ")

time1 = t.now()

def prefix_function(text):
    n = len(text)
    pi = [0] * n
    for i in range(1, n):
        j = pi[i-1]
        while j > 0 and text[j] != text[i]:
            j = pi[j-1]
        if text[i] == text[j]:
            j += 1
        pi[i] = j
    return pi

def KMP_algorithm(text, sub_text, start_index, button):
    start_index = 0
    if button == 'off':
        text = text.lower()
        sub_text = sub.text.lower()
    j = 0
    pi = prefix_function(sub_text)
    for i in range(start_index, len(text)):
        while j > 0 and text[i] != sub_text[j]:
            j = pi[j-1]
        if text[i] == sub_text[j]:
            j += 1
        if j >= len(sub_text):
            return "Подстрока найдена"
    return "Подстрока не найдена"

print(KMP_algorithm(text, sub_text, 0, button))

time2 = t.now()
print("--- {0} ms ---".format((time2 - time1).microseconds))

Cтрока: It's the thought of being young When your heart's just like a drum Beating louder with no way to guard it When it all seems like it's wrong Just sing along to Elton John And to that feeling, we're just getting started When the nights get colder And the rhythms got you falling behind Just dream about that moment When you look yourself right in the eye, eye, eye Then you say I wanna dance The music's got me going Ain't nothing that can stop how we move, yeah Let's break our plans And live just like we're golden And roll in like we're dancing fools We don't need to worry 'Cause when we fall, we know how to land Don't need to talk the talk, just walk the walk tonight 'Cause we don't need permission to dance
Подстрока: Elton John
Выбор опции чувствительности к регистру -  on or off on
Подстрока найдена
--- 1289 ms ---


#### Упрощенный алгоритм Бойера-Мура

In [11]:
text = input("Строка: ")
sub_text = input("Подстрока: ")
button = input("Выбор опции чувствительности к регистрн -  on or off:  ")

time1 = t.now()

def dic(sub_text):
    S = set()  
    M = len(sub_text) 
    dictionary = defaultdict()       
    for i in range(M-2, -1, -1): 
        if sub_text[i] not in S:        
            dictionary[sub_text[i]] = M-i-1
            S.add(sub_text[i])
    if sub_text[M-1] not in S:     
        dictionary[sub_text[M-1]] = M
    dictionary['*'] = M              
    return dictionary

def BM_algorithm(text, sub_text, button):
    if button == 'off':
        text = text.lower()
        sub_text = sub_text.lower()
    dictionary = dic(sub_text)
    N = len(text)
    M = len(sub_text)
    if N >= M:
        i = M-1       
        while(i < N):
            k = 0
            j = 0
            flBreak = False
            for j in range(M-1, -1, -1):
                if text[i-k] != sub_text[j]:
                    if j == M-1:
                        off = dictionary[text[i]] if dictionary.get(text[i], False) else dictionary['*']  
                    else:
                        off = dictionary[sub_text[j]]   
                    i += off    
                    flBreak = True 
                    break
                k += 1          
            if not flBreak:         
                return "Подстрока найдена"
                break
        else:
            return "Подстрока не найдена"
    else:
        return "Подстрока не найдена"
    

print(BM_algorithm(text, sub_text,button))

time2 = t.now()
print("--- {0} ms ---".format((time2 - time1).microseconds))

Строка: It's the thought of being young When your heart's just like a drum Beating louder with no way to guard it When it all seems like it's wrong Just sing along to Elton John And to that feeling, we're just getting started When the nights get colder And the rhythms got you falling behind Just dream about that moment When you look yourself right in the eye, eye, eye Then you say I wanna dance The music's got me going Ain't nothing that can stop how we move, yeah Let's break our plans And live just like we're golden And roll in like we're dancing fools We don't need to worry 'Cause when we fall, we know how to land Don't need to talk the talk, just walk the walk tonight 'Cause we don't need permission to dance
Подстрока: elton john
Выбор опции чувствительности к регистрн -  on or off:  on
Подстрока не найдена
--- 1543 ms ---


#### Стандартная функция поиска

In [13]:
text = input('Строка: ')
sub_text = input('Строка: ')
button = input("Выбор опции чувствительности к регистрн -  on or off: ")

time1 = t.now()

if button == 'off':
    text = text.lower()
    sub_text = sub_text.lower()
if text.find(sub_text) != -1:                     
    print('Подстрока найдена')
else:
    print ('Подстрока не найдена')

time2 = t.now()
print("--- {0} ms ---".format((time2 - time1).microseconds))

Строка: It's the thought of being young When your heart's just like a drum Beating louder with no way to guard it When it all seems like it's wrong Just sing along to Elton John And to that feeling, we're just getting started When the nights get colder And the rhythms got you falling behind Just dream about that moment When you look yourself right in the eye, eye, eye Then you say I wanna dance The music's got me going Ain't nothing that can stop how we move, yeah Let's break our plans And live just like we're golden And roll in like we're dancing fools We don't need to worry 'Cause when we fall, we know how to land Don't need to talk the talk, just walk the walk tonight 'Cause we don't need permission to dance
Строка: Elton john
Выбор опции чувствительности к регистрн -  on or off: off
Подстрока найдена
--- 314 ms ---


### Задание №2
Написать программу, определяющую, является ли данное
расположение «решаемым», то есть можно ли из него за конечное число
шагов перейти к правильному. Если это возможно, то необходимо найти хотя
бы одно решение - последовательность движений, после которой числа будут
расположены в правильном порядке.
#### Входные данные: массив чисел, представляющий собой расстановку в
Порядке «слева направо, сверху вниз». Число 0 обозначает пустое поле.
Например, массив [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 0] представляет
собой «решенную» позицию элементов.
#### Выходные данные: если решения нет, то функция должна вернуть
Пустой массив []. Если решение есть, то необходимо представить решение —
для каждого шага записывается номер передвигаемого на данном шаге
элемента. 

In [1]:
from queue import PriorityQueue

N = 4

def moves(position):
    blank = position.index(0)
    i, j = divmod(blank, N)
    offsets = []
    if i > 0: offsets.append(-N)     
    if i < N - 1: offsets.append(N) 
    if j > 0: offsets.append(-1)     
    if j < N - 1: offsets.append(1)  
    for offset in offsets:
        swap = blank + offset
        yield tuple(position[swap] if x==blank else position[blank] if x==swap else position[x] for x in range(N*N))

def parity(permutation):
    seen, cycles = set(), 0
    for i in permutation:
        if i not in seen:
            cycles += 1
            while i not in seen:
                seen.add(i)
                i = permutation[i]
    return (cycles + len(permutation)) % 2


class Position:
    def __init__(self, position, start_distance):
        self.position = position
        self.start_distance = start_distance

    def __lt__(self, other):
        return self.start_distance < other.start_distance

    def __str__(self):
        return '\n'.join((N*'{:3}').format(*[i%(N*N) for i in self.position[i:]]) 
for i in range(0, N*N, N))

SOLVED = (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 0)
start = [1, 2, 3, 4, 5, 6, 7, 8, 13, 9, 11, 12, 10, 14, 15, 0]

if parity(start) == 0:
    print('[]')
else:
    start = tuple(start)

    p = Position(start, 0)

    print(p)
    print()

    candidates = PriorityQueue()
    candidates.put(p)

    visited = set([p])

    came_from = {p.position: None}
    
    while p.position != SOLVED:
        p = candidates.get()
        for k in moves(p.position):
            if k not in visited:
                candidates.put(Position(k, p.start_distance + 1))
                came_from[k] = p
                visited.add(k)

    path = []
    prev = p
    while p.position != start:
        p = came_from[p.position]
        number = p.position[prev.position.index(0)]
        path.append(number)
        prev = p
    path.reverse()

    print(path)

  1  2  3  4
  5  6  7  8
 13  9 11 12
 10 14 15  0

[15, 14, 10, 13, 9, 10, 14, 15]


### Вывод

##### В результате оценки времени работы алгоритмов поиска подстроки в строке, для текста состоящего из 712 символов, были получены следующие результаты:
     1) Алгоритм Кнута-Морриса-Пратта - 1289 ms
     2) Упрощенный алгоритм Бойера-Мура - 1543 ms
     3) Стандартная функция поиска - 314 ms
     Стандартная функция поиска оказалась наиболее быстрой по времени; 
     Наиболее быстрая по времни, за исключением стандартной функции, - алгоритм Кнута-Морриса-Пратта.
     Наиболее длительная - упрощенный алгоритм Бойера-Мура.